polar-core/src/data_filtering.rs

```diff
@@ -0,0 +1,799 @@
+    // and if they had aliases they'd be collapsed by now, so it should be an error
+    // if foo.name = "steve" and foo.name = "gabe".
+    let mut eq_values = HashMap::new();
+    'values: for (var, value) in var_info.eq_values {
+        for (id, set) in &mut variables {
+            if set.contains(&var) {
+                // @TODO(steve): If we already have a value for it make sure they match don't just
+                // overwrite it.
+                eq_values.insert(id.clone(), value);
+                continue 'values;
+            }
+        }
+        // Create new variable if we didn't find one.
+        let new_id = get_id();
+        let mut new_set = HashSet::new();
+        new_set.insert(var.clone());
+        variables.insert(new_id.clone(), new_set);
+        eq_values.insert(new_id, value);
+    }
+
+    let mut contained_values = HashMap::new();
+    'contained_values: for (value, var) in var_info.contained_values {
+        for (id, set) in &mut variables {
+            if set.contains(&var) {
+                contained_values
+                    .entry(id.clone())
+                    .or_insert_with(HashSet::new)
+                    .insert(value);
+                continue 'contained_values;
+            }
+        }
+        // Create new variable if we didn't find one.
+        let new_id = get_id();
+        let mut new_set = HashSet::new();
+        new_set.insert(var.clone());
+        variables.insert(new_id.clone(), new_set);
+        let mut new_val_set = HashSet::new();
+        new_val_set.insert(value);
+        contained_values.insert(new_id, new_val_set);
+    }
+
+    let mut types = HashMap::new();
+    'types: for (var, typ) in var_info.types {
+        for (id, set) in &mut variables {
+            if set.contains(&var) {
+                // @TODO(steve): If we already have a type for it make sure they match don't just
+                // overwrite it.
+                types.insert(id.clone(), typ);
+                continue 'types;
+            }
+        }
+        // Create new variable if we didn't find one.
+        let new_id = get_id();
+        let mut new_set = HashSet::new();
+        new_set.insert(var.clone());
+        variables.insert(new_id.clone(), new_set);
+        types.insert(new_id, typ);
+    }
+
+    let mut this_id = String::new();
+    for (id, set) in &variables {
+        if set.contains(&Symbol::new("_this")) {
+            this_id = id.clone()
+        }
+    }
+
+    Vars {
+        variables,
+        field_relationships,
+        in_relationships,
+        eq_values,
+        contained_values,
+        types,
+        this_id,
+    }
+}
+
+fn constrain_vars(types: &Types, vars: &Vars, this_type: &str) -> ResultSet {
+    let mut result_set = ResultSet {
+        requests: HashMap::new(),
+        resolve_order: vec![],
+        result_id: vars.this_id.clone(),
+    };
+    constrain_var(&mut result_set, types, vars, &vars.this_id, this_type);
+    result_set
```

Comment:
How are we actually checking that everything in Vars is converted into a constraint?

For example if I wrote something like `x.bar = x.y.foo`? My undersatnding is that isn't supported right now, but I think it would just silently not be checked. Is that right?